### PR TITLE
Changed add_token.py for issue #4526 

### LINKED ISSUE #4526: microk8s add-node --format=short outputs the join command twice
#### Summary
The output appears to be duplicated.
```
$ microk8s add-node --format=short
microk8s join redacted:25000/redacted/redacted
microk8s join redacted:25000/redacted/redacted
```

#### What Should Happen Instead?
One line should be output instead of two.

#### Reproduction Steps

1. Install MicroK8s v1.28.9 revision 6750
2. Run `microk8s add-node --format=short`

#### Introspection Report

n/a

#### Are you interested in contributing with a fix?
Not at this time.

### PR DESCRIPTION
**Pull request by ShrishtiKarkera**
#### Changes

/microk8s/scripts/wrappers/add_token.py

#### Testing

The change was tested with a unittest in tests/unit/test_addtoken.py
Tested using mock print assertion with builtins.print

#### Checklist

* [Yes] Read the [contributions](https://github.com/canonical/microk8s/blob/master/CONTRIBUTING.md) page.
* [Yes] Submitted the [CLA form](https://ubuntu.com/legal/contributors/agreement), if you are a first time contributor.
* [Yes] The introduced changes are covered by unit and/or integration tests.

### CODE REVIEW

File: scripts/wrappers/add_token.py [hunk 1]
```diff
@@ -120,10 +120,15 @@ def print_yaml(token, check):
...
+    default_addr = str(default_ip) + ":" + str(port) + "/" + str(token) + "/" + str(check)
+    addr_set = set()
+    addr_set.add(default_addr)
     print(f"microk8s join {default_ip}:{port}/{token}/{check}")
     for ip in all_ips:
-        print(f"microk8s join {ip}:{port}/{token}/{check}")
+        addr = str(ip) + ":" + str(port) + "/" + str(token) + "/" + str(check)
+        if addr not in addr_set:
+            print(f"microk8s join {ip}:{port}/{token}/{check}")
+            addr_set.add(addr)
```
Comment (neoaggelos) [inline scripts/wrappers/add_token.py:131]: This is creating a set of IPs we have already "seen". However, the ask here is to simply avoid repeating the default_ip. It should be simpler if you simply check for `ip == default_ip` in the for loop, right?

File: tests/unit/test_addtoken.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from unittest import mock, TestCase
+from add_token import print_short
+
+class test_at(TestCase):
+    @mock.patch('builtins.print')
+    @mock.patch('add_token.get_network_info', return_value=['10.23.53.54', ['10.23.53.54'], str(32)])
+    def test_add_token(self, addtoken, mock_print):
+        # self.assertEqual(print_short('t', 'c'), 2)
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:8]: is this leftover from debugging? do we need it?

File: tests/unit/test_addtoken.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from unittest import mock, TestCase
+from add_token import print_short
+
+class test_at(TestCase):
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:4]: test classes are typically named with camel-case: `TestSomething`. Please use a meaningful name for what we test here.

File: tests/unit/test_addtoken.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    @mock.patch('builtins.print')
+    @mock.patch('add_token.get_network_info', return_value=['10.23.53.54', ['10.23.53.54'], str(32)])
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:6]: let's be consistent and use `'32'` as well, instead of `str(32)`

File: tests/unit/test_addtoken.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    def test_add_token(self, addtoken, mock_print):
+        # self.assertEqual(print_short('t', 'c'), 2)
+        print_short('t', 'c')
+        mock_print.assert_called_with('microk8s join 10.23.53.54:32/t/c')
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:10]: this is great, but only checks one trivial case. could you add (at least one) extra test with more IPs to be printed, where we verify the result?

Ideally, the tests should be a useful signal to know that we don't break the function in all the cases where it is use. If we only test one trivial input, then that signal might not be very trustworthy.

**Review by neoaggelos — COMMENTED** [review]
Thanks a lot @ShrishtiKarkera! This is going in the right direction, and solves the issue at hand. However, there are things that can be improved, please have a look at the comments.

File: scripts/wrappers/add_token.py (hunk 1 above)
Comment (ShrishtiKarkera) [reply scripts/wrappers/add_token.py:131]: Alright!
I was thinking if the there were duplicate ips in the all_ip list apart from default ip but its much simpler if the default ip is the only problem.
Thanks!

File: tests/unit/test_addtoken.py (hunk 3 above)
Comment (ShrishtiKarkera) [reply tests/unit/test_addtoken.py:4]: My bad, test_at meant test_addtoken but I'll be more clearer.
Thanks

File: tests/unit/test_addtoken.py (hunk 4 above)
Comment (ShrishtiKarkera) [reply tests/unit/test_addtoken.py:6]: Got it.

File: tests/unit/test_addtoken.py (hunk 5 above)
Comment (ShrishtiKarkera) [reply tests/unit/test_addtoken.py:9]: Sure, I'll do that.

**Review by ShrishtiKarkera — COMMENTED** [review]
Changed add_token.py to just check for the default_ip case.
Added another test case with multiple ip addresses.

File: tests/unit/test_addtoken.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    @mock.patch('add_token.get_network_info', return_value=['10.23.53.54', ['10.23.53.54'], '32'])
+    def test_single(self, single, mock_print):
+        print_short('t', 'c')
+        mock_print.assert_called_with('microk8s join 10.23.53.54:32/t/c')
+    
+    @mock.patch('builtins.print')
+    @mock.patch('add_token.get_network_info', return_value=['d_ip', ['ip1', 'ip2', 'd_ip'], '4'])
+    def test_multiple(self, multiple, mock_print):
+        print_short('t', 'c')
+        mock_print.assert_called_with('microk8s join ip2:4/t/c')
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:15]: this works, but does not tell the whole story. please make sure that you test all print invocations are as expected, not only the last one.

File: tests/unit/test_addtoken.py (hunk 5 above)
Comment (neoaggelos) [reply tests/unit/test_addtoken.py:9]: Thanks!

**Review by neoaggelos — COMMENTED** [review]
Looks nicer! Did a second pass

File: tests/unit/test_addtoken.py (hunk 6 above)
Comment (ShrishtiKarkera) [reply tests/unit/test_addtoken.py:15]: Sure!

**Review by ShrishtiKarkera — COMMENTED** [review]
I couldn't use the inbuilt assertion as it has a limit so I used capsys to check for all output lines.
Please let me know if anything else is required to be done.

File: tests/unit/test_addtoken.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    
+def test_multiple(capsys):
+    with mock.patch('add_token.get_network_info', return_value=['d_ip', ['ip1', 'ip2', 'd_ip'], '4']):
+        print_short('t', 'c')
+        captured = capsys.readouterr()
+        all_outputs = captured.out.strip().split('\n')
+        all_outputs[0] == f"microk8s join d_ip:4/t/c"
+        assert all_outputs[1] == f"microk8s join ip1:4/t/c"
+        assert all_outputs[2] == f"microk8s join ip2:4/t/c"
+        assert len(all_outputs) == 3
```
Comment (neoaggelos) [inline tests/unit/test_addtoken.py:20]: this can be a single `assert all_outputs == [cmd1, cmd2, cmd3]`

Comment (ShrishtiKarkera) [reply tests/unit/test_addtoken.py:20]: Okay

**Review by ShrishtiKarkera — COMMENTED** [review]
Removed multiple assertions and added a single assert statement

**Review by ShrishtiKarkera — COMMENTED** [review]
Checked for cla and lint

**Review by ShrishtiKarkera — COMMENTED** [review]
checked for lint and cla

**Review by ShrishtiKarkera — COMMENTED** [review]
Lint passes, checking for cla

**Review by ShrishtiKarkera — COMMENTED** [review]
checking for cla

**Review by ShrishtiKarkera — COMMENTED** [review]
Final check

**Review by ShrishtiKarkera — COMMENTED** [review]
Checked all tests

File: scripts/wrappers/join.py [hunk 8]
```diff
@@ -594,10 +594,13 @@ def update_dqlite(cluster_cert, cluster_key, voters, host):
...
-        data = yaml.safe_load(f)
-    if "Address" in data:
-        port = data["Address"].rsplit(":")[-1]
+    try:
+        with open("{}/info.yaml".format(cluster_backup_dir)) as f:
+            data = yaml.safe_load(f)
+        if "Address" in data:
+            port = data["Address"].rsplit(":")[-1]
+    except OSError:
+        pass
```
Comment (neoaggelos) [inline scripts/wrappers/join.py:604]: this is unrelated to the changes of this PR, please revert

**Review by neoaggelos — COMMENTED** [review]
Please rebase and cleanup the commit history, then we should be good to merge!

File: scripts/wrappers/join.py (hunk 8 above)
Comment (ShrishtiKarkera) [reply scripts/wrappers/join.py:604]: Sure

**Review by ShrishtiKarkera — COMMENTED** [review]
Removed the join.py commit from the commit history

**Review by ShrishtiKarkera — COMMENTED** [review]
Squashed all commits into one

**Review by neoaggelos — APPROVED** [review]
